src/ck-ldb/TreeLB.C

```diff
@@ -0,0 +1,552 @@
+  // a check here is needed because PE might have been sending chares/tokens or
+  // receiving them even before receiving the decision msg (with the final tally)
+  // from the parent. So, exchange might have already completed
+  checkLoadExchanged(level);
+}
+
+// order transfer of load token to dest
+void TreeLB::transferLoadToken(TreeLBMessage* transferMsg, int level)
+{
+  // TODO this is a simplified implementation where the first level where tokens are
+  // needed needs to be able to generate the tokens (so, token requests cannot be
+  // propagated down the tree). There is a more advanced implementation in charm4py
+  if (logic[level]->makesTokens())
+  {
+    // one token set goes to one destination
+    std::vector<TreeLBMessage*> token_sets;
+    std::vector<int> destinations;
+    load_sent[level] += logic[level]->getTokenSets(transferMsg, token_sets, destinations);
+    for (size_t i = 0; i < token_sets.size(); i++)
+    {
+      token_sets[i]->level = level;
+      thisProxy[destinations[i]].recvLoadTokens((CkMessage*)token_sets[i]);
+    }
+  }
+  else
+  {
+    // implemented in python. see code
+    CkAbort("TreeLB::transferLoadToken - NOT IMPLEMENTED\n");
+  }
+}
+
+void TreeLB::recvLoadTokens(CkMessage* tokens)
+{
+  TreeLBMessage* token_set = (TreeLBMessage*)tokens;
+  int level = token_set->level;
+#if DEBUG__TREE_LB_L3
+  fprintf(stderr, "[%d] Received load token, level=%d\n", CkMyPe(), level);
+#endif
+  int load = logic[level]->tokensReceived(token_set);
+  load_received[level] += load;
+  checkLoadExchanged(level);
+}
+
+void TreeLB::objMovedIn(bool waitBarrier)
+{
+  if (!waitBarrier) CkAbort("TreeLB future migrates not supported\n");
+
+  // fprintf(stderr, "[%d] TreeLB::objMovedIn\n", CkMyPe());
+
+  int level = 0;
+  CkAssert(numLevels > 0 && awaitingLB[level]);
+  load_received[level] += 1;
+  checkLoadExchanged(level);
+}
+
+void TreeLB::migrateObjects(const IDM& mig_order)
+{
+  int level = 0;
+  int sent = logic[level]->migrateObjects(mig_order.data.at(CkMyPe()));
+  load_sent[level] += sent;
+#if DEBUG__TREE_LB_L2
+  fprintf(stderr, "[%d] Received IDM order, sent=%d\n", CkMyPe(), sent);
+#endif
+  checkLoadExchanged(level);
+}
+
+void TreeLB::lb_done()
+{
+  // fprintf(stderr, "[%d] lb_done step %d lb_time=%f\n", CkMyPe(), lbmgr->step(),
+  // CkWallTimer() - startTime);
+
+  // TODO LBManager should do all of this, including global syncResume ******
+  // Currently, TreeLB does syncResume by setting barrier_after_lb=true
+
+  // clear load stats
+  lbmgr->ClearLoads();
+
+  if (CkMyPe() == 0 && _lb_args.debug() > 0)
+  {
+    CkPrintf("--------- Finished LB step %d ---------\n", lbmgr->step());
+  }
+
+  // Increment to next step
+  lbmgr->incStep();
+
+  LBManager::Object()
+      ->MigrationDone();  // call registered callbacks. not sure what this is for
+
+  // again, TreeLB shouldn't be doing these things. it should just notify LBManager
```

Comment:
Do we have an open issue or issues for the various todo's mentioned in comments here?